### PR TITLE
Read poi and expectation directly from `output_filename` to accelerate `NeymanConstructor`

### PR DESCRIPTION
**Pull request by dachengx**
We do not need to initialize runner for each **{**nominal_values, **generate_values}, but we can read the poi directly from toymc files `output_filename`.

This was initially an idea from @hammannr, https://github.com/XENONnT/alea/pull/100#issuecomment-1784817837.

0. Save expectation_values to metadata
1. First figure out whether it is single `output_filename` mode or multiple(batch) `output_filename` mode
2. Read metadata from the list of `output_filename` and make sure that the metadata is the same in the list
3. Take the first entry of the metadata list as metadata
4. Assert it contains poi, which is guaranteed by https://github.com/XENONnT/alea/blob/80bd030205ed299fd26aedaddfdc31f3ee5849bb/alea/runner.py#L340
5. Read poi and expectation_values from metadata
6. Read poi_expectation from expectation_values

This PR is superseding https://github.com/XENONnT/alea/pull/103.

### CODE REVIEW

File: alea/models/blueice_extended_model.py [hunk 1]
```diff
@@ -138,11 +139,31 @@ def get_source_name_list(self, likelihood_name: str) -> list:
...
+
+        Returns:
+            set: set of source names.
+
+        """
+        source_name = set(
+            itertools.chain.from_iterable([ll.source_name_list for ll in self.likelihood_list[:-1]])
+        )
+        return source_name
+
```
Comment (hammannr) [inline alea/models/blueice_extended_model.py:156]: This is not used anywhere. I would remove it

File: alea/models/blueice_extended_model.py [hunk 2]
```diff
@@ -138,11 +139,31 @@ def get_source_name_list(self, likelihood_name: str) -> list:
...
     @property
     def likelihood_list(self) -> List:
         """Return a list of likelihood terms."""
         return self._likelihood.likelihood_list
 
+    @property
+    def likelihood_parameters(self) -> List:
+        """Return a list of likelihood parameters."""
+        return self._likelihood.likelihood_parameters
+
```
Comment (hammannr) [inline alea/models/blueice_extended_model.py:166]: I don't mind it but also this is not related to this PR, is it?

File: alea/models/blueice_extended_model.py (hunk 1 above)
Comment (hammannr) [reply alea/models/blueice_extended_model.py:156]: also, it doesn't take `likelihood_name` and I think if at all it should be `get_all_source_names`.

File: alea/submitters/local.py [hunk 3]
```diff
@@ -156,19 +158,82 @@ def submit(
...
+                else:
+                    warnings.warn(
+                        "Can not find the poi_expectation in the generate_values "
+                        f"{generate_values}, so will not check if the poi_expectation "
+                        "are consistent!"
+                    )
+                    poi_expectation = _poi_expectation
+
+            # read the likelihood ratio
+            results = toyfiles_to_numpy(output_filename)
```
Comment (hammannr) [inline alea/submitters/local.py:236]: Can we test some of this?

Comment (hammannr) [reply alea/submitters/local.py:236]: In general, this method has become pretty lengthy. Do you think it could make sense to break out some functionality into smaller functions that you can then call in `submit`? This makes it easier to test and to read 😊

**Review by hammannr — COMMENTED** [review]
I didn't go through it line-by-line yet but I think it looks good!
I already added a few comments and I'm now running a small toy study to see if everything works as expected. I'll let you know as soon as it's done 😊

File: alea/models/blueice_extended_model.py (hunk 1 above)
Comment (dachengx) [reply alea/models/blueice_extended_model.py:156]: Suspended by https://github.com/XENONnT/alea/pull/111

File: alea/submitters/local.py (hunk 3 above)
Comment (dachengx) [reply alea/submitters/local.py]: The functions are split. I have no idea currently how to test the code of `NeymanConstructor` because then we need to save the output files or run the configs when testing. That can be time consuming.

Comment (hammannr) [reply alea/submitters/local.py]: You're right. Maybe we can write and load some trivial example or something? But it's not that straightforward, I agree.

**Review by hammannr — APPROVED** [review]
The split version is now much more readable -- I like it! I think we should think about unittests for the `NeymanConstructor` but I'd suggest to merge this now without them.

I ran the code with some examples and the results were sensible. Thanks a lot @dachengx for this improvement 😊
